user/user_app.go

```diff
@@ -0,0 +1,106 @@
+// are allowed, using user-defined criteria. Will be one of: contracts, methods, origins, services, user_agent
+type AllowlistType string
+
+// TODO_DISCUSS: which allowlist types are needed in PATH?
+const (
+	// AllowlistTypeContracts allows the UserApp to specify which blockchain contracts are allowed to be passed in service requests.
+	AllowlistTypeContracts AllowlistType = "contracts"
+	// AllowlistTypeMethods allows the UserApp to specify which methods are allowed to be used for service requests.
+	AllowlistTypeMethods AllowlistType = "methods"
+	// AllowlistTypeOrigins allows the UserApp to specify which request origins are allowed to make service requests.
+	AllowlistTypeOrigins AllowlistType = "origins"
+	// AllowlistTypeServices allows the UserApp to specify which services (eg. blockchains relay IDs) are allowed to make service requests.
+	AllowlistTypeServices AllowlistType = "services"
+	// AllowlistTypeUserAgents allows the UserApp to specify which user agents are allowed to make service requests.
+	AllowlistTypeUserAgents AllowlistType = "user_agents"
+)
+
+// A UserApp represents a user's application, which has two primary functions:
+// 1. Identifying which user is making a service request, to enabled metering, billing, and rate limiting
+// 2. Allowing configuration of user-specific settings, such as secret key authentication, allowlists, etc.
+//
+// A UserApp is associated to a single Account. An Account can have multiple UserApps and one PlanType.
+// Settings related to service requests, such as enforcing secret key authentication, are configured per UserApp.
+// Pricing plan and team management settings are configured per Account.
+type UserApp struct {
+
+	// The unique identifier for a user's application, which must be passed
+	// as the last path segment of the service request URL. eg `/v1/{user_app_id}`
+	// This is used to identify the UserApp when making a service request.
+	ID UserAppID
+	// The unique identifier for a UserApp's Account, used for metering & billing purposes.
+	AccountID AccountID
+	// The plan type for a UserApp's Account, which identifies the pricing plan for the Account.
+	PlanType PlanType
+	// The throughput rate limit for UserApp, in requests per second.
+	RateLimitThroughput int
+
+	// The secret key for UserApp. If SecretKeyRequired is true, the secret key
+	// must be passed as the `Authentication` header in service requests.
+	SecretKey string
+	// Whether the UserApp requires a secret key for authentication.
+	// If not set, the UserApp does not require a secret key for authentication.
+	SecretKeyRequired bool
+
+	// The allowlists for the UserApp, which allow the UserApp to specify certain security settings,
+	// for example, which contracts, methods, origins, services, or user agents are allowed to make
+	// service requests through the Gateway with the UserApp.
+	Allowlists map[AllowlistType]map[string]struct{}
```

Comment:
Can we remove all of this (Auth/Secret Key, AllowList, etc...) from this PR?

Seems like there is a lot of discussion (naming, how we want to implement it, structure, etc...) that's outside the scope of this PR.

---

Rationale:
1. Auth is a solved problem. I want to understand if we should pick an opinionated solution (e.g. clerk, auth0, etc) and how we integrate it.
2. Allow/block lists is very critical but it looks like we're just tagging along poral-middleware techdebt. At the very least, this should be a separate file/structure/etc. 

Again, both of these can be multiple days of thought/review and I don't see the urgency in having a port of the portal-middleware solution here this week creating TECHDEBT so soon.

---

Note: As an engineer, if you're worried about the "but we already implemented it", what I would do is:
1. Remove it
2. `git diff . >> /tmp/tmp.diff`
3. Create a new GitHub issue and upload `tmp.diff`
4. Use that as a structure point in the followup

Reply:
Regarding Auth I'm fine with moving it to (https://github.com/buildwithgrove/path/pull/18) since that's where it's actually used for the first time. That said, I don't really see the benefit to removing it only to put it back two PRs down. Your call.

> 1. Auth is a solved problem. I want to understand if we should pick an opinionated solution (e.g. clerk, auth0, etc) and how we integrate it.

I would strongly suggest we don't introduce an external dependency for this. As I mentioned above, it's really just a very simple API key check: does the request `Authorization` header contain the correct `SecretKey`? Yes or no. Adding something with auth0 is way overkill for this IMO.

_(Again, we can rename `SecretKey` to `APIKey` for clarity; I'm leaning towards this at this point.)_

> 2. Allow/block lists is very critical but it looks like we're just tagging along poral-middleware techdebt. At the very least, this should be a separate file/structure/etc.

This was actually my attempt to clean up/simplify the way we handle Allowlists in the the Portal. 

For example this is the struct we use currently:
```go
struct {
		Origins     map[Origin]struct{}                    `json:"origins,omitempty"`
		UserAgents  map[UserAgent]struct{}                 `json:"userAgents,omitempty"`
		Blockchains map[RelayChainID]struct{}              `json:"blockchains,omitempty"`
		Contracts   map[RelayChainID]map[Contract]struct{} `json:"contracts,omitempty"`
		Methods     map[RelayChainID]map[Method]struct{}   `json:"methods,omitempty"`
	}
```

That said, I'm fine with removing it for now if you like, but with the understanding that this is something we will be using very soon and it will need to be added back when we do.

Reply:
I've removed all references to `Allowlists` for now. We can discuss the implementation when we get to the point of adding them back.

I've kept `Auth`.